### ISSUE TITLE
Custom encoder/decoder avoiding Protobuf

### ISSUE THREAD

**Issue by prologic** [prologic/bitcask#42]
Off the back of the discussion in #33 was the suggestion that we build our own custom encoder/decoder like the original paper had set for laying out the encoded entries on disk:

```
{uint32 key size}{uint64 value size}{key}{value}{checksum}
^--------------12 bytes-----------^
```

In my own words:

> Re use of Protobuf -- You are right! I admit I was a bit lazy and didn't go to all the effort of hand-writing a custom encoder/decoder -- I suspect it may be more efficient/faster if we did! /cc @jsign :D -- Would you mind filding a separate issue on this so we can explore this more? (I have to admit I kind of detest how much Protobuf adds to a project and resulting binary's weight :/)

When an entry is written to file, three things are stored: the key, the value, and `checksum(value)`.

Currently protobuf is used to serialise this data into an entry before it is written to file.

https://github.com/prologic/bitcask/blob/8b684b635de2367c1776e57046a021752e731283/internal/entry.go#L9-L17

The data being stored is relatively simple. There is one key key and one value, each of which are variable length. Thus a simple scheme where the length of the key and value is stored at the start followed by the key and the value and then the fixed-size checksum at the end.

This would likely be faster than using protobuf with a lot less overhead. We can also apply our own context-specific optimisations that protobuf would be blind/apathetic towards.

**Comment by awnumar** [comment]
Something to note is that after this change it will not be possible to open existing databases.

**Comment by prologic** [comment]
This is true. It would be possible to create a "conversion" tool if anyone really needed that.

**Comment by prologic** [comment]
@jsign You might like to take this one on? :)

**Comment by jsign** [comment]
Sure, I should have some more days with extra time... I'll do a WIP PR when progress exists.

**Comment by jsign** [comment]
Well, I have some preliminary interesting results. I think I took this at the possible limit (hopefully, correctly). :rocket: 
_I did a small trick to avoid the `string` conversion for the `key`, but it won't be necessary when changing the key to `byte[]` arrives, so the improvement holds._

An interesting consequence of doing this optimization: this implementation doesn't need any mempools. :) I tried to be careful with helping the compiler decide not escaping slices to the heap when possible.

Bench results:
```
goos: linux
goarch: amd64
pkg: github.com/prologic/bitcask
BenchmarkGet/128B-8              2000000               937 ns/op         136.54 MB/s         160 B/op          1 allocs/op
BenchmarkGet/256B-8              2000000               889 ns/op         287.73 MB/s         288 B/op          1 allocs/op
BenchmarkGet/512B-8              1000000              1051 ns/op         486.93 MB/s         576 B/op          1 allocs/op
BenchmarkGet/1K-8                1000000              1249 ns/op         819.26 MB/s        1152 B/op          1 allocs/op
BenchmarkGet/2K-8                1000000              1629 ns/op        1256.74 MB/s        2304 B/op          1 allocs/op
BenchmarkGet/4K-8                 500000              2492 ns/op        1643.57 MB/s        4864 B/op          1 allocs/op
BenchmarkGet/8K-8                 500000              3844 ns/op        2130.90 MB/s        9472 B/op          1 allocs/op
BenchmarkGet/16K-8                200000              6965 ns/op        2352.30 MB/s       18432 B/op          1 allocs/op
BenchmarkGet/32K-8                100000             14140 ns/op        2317.39 MB/s       40960 B/op          1 allocs/op
BenchmarkPut/128B-8               500000              4016 ns/op          31.87 MB/s         201 B/op          6 allocs/op
BenchmarkPut/256B-8               300000              5209 ns/op          49.14 MB/s         202 B/op          6 allocs/op
BenchmarkPut/1K-8                 100000             12305 ns/op          83.21 MB/s         210 B/op          6 allocs/op
BenchmarkPut/2K-8                 100000             19231 ns/op         106.49 MB/s         221 B/op          6 allocs/op
BenchmarkPut/4K-8                  50000             37505 ns/op         109.21 MB/s         241 B/op          6 allocs/op
BenchmarkPut/8K-8                  20000             69770 ns/op         117.41 MB/s         288 B/op          6 allocs/op
BenchmarkPut/16K-8                 10000            137953 ns/op         118.76 MB/s         364 B/op          6 allocs/op
BenchmarkPut/32K-8                  5000            264327 ns/op         123.97 MB/s         531 B/op          7 allocs/op
BenchmarkScan-8                  1000000              1547 ns/op             408 B/op         16 allocs/op
```

Bench of current `master`:
```
goos: linux
goarch: amd64
pkg: github.com/prologic/bitcask
BenchmarkGet/128B-8              1000000              1200 ns/op         106.58 MB/s         387 B/op          4 allocs/op
BenchmarkGet/128BWithPool-8      1000000              1302 ns/op          98.28 MB/s         227 B/op          3 allocs/op
BenchmarkGet/256B-8              1000000              1222 ns/op         209.34 MB/s         643 B/op          4 allocs/op
BenchmarkGet/256BWithPool-8      1000000              1254 ns/op         203.99 MB/s         355 B/op          3 allocs/op
BenchmarkGet/512B-8              1000000              1411 ns/op         362.80 MB/s        1187 B/op          4 allocs/op
BenchmarkGet/512BWithPool-8      1000000              1382 ns/op         370.25 MB/s         611 B/op          3 allocs/op
BenchmarkGet/1K-8                1000000              1775 ns/op         576.72 MB/s        2275 B/op          4 allocs/op
BenchmarkGet/1KWithPool-8        1000000              1616 ns/op         633.59 MB/s        1123 B/op          3 allocs/op
BenchmarkGet/2K-8                 500000              2499 ns/op         819.35 MB/s        4451 B/op          4 allocs/op
BenchmarkGet/2KWithPool-8        1000000              2072 ns/op         988.27 MB/s        2147 B/op          3 allocs/op
BenchmarkGet/4K-8                 500000              3883 ns/op        1054.79 MB/s        9059 B/op          4 allocs/op
BenchmarkGet/4KWithPool-8         500000              2987 ns/op        1371.09 MB/s        4195 B/op          3 allocs/op
BenchmarkGet/8K-8                 300000              6588 ns/op        1243.32 MB/s       17763 B/op          4 allocs/op
BenchmarkGet/8KWithPool-8         300000              4790 ns/op        1710.14 MB/s        8291 B/op          3 allocs/op
BenchmarkGet/16K-8                200000             11548 ns/op        1418.71 MB/s       34915 B/op          4 allocs/op
BenchmarkGet/16KWithPool-8        200000              8036 ns/op        2038.63 MB/s       16483 B/op          3 allocs/op
BenchmarkGet/32K-8                100000             22002 ns/op        1489.26 MB/s       73827 B/op          4 allocs/op
BenchmarkGet/32KWithPool-8        100000             15304 ns/op        2141.12 MB/s       32867 B/op          3 allocs/op
BenchmarkPut/128B-8               300000              4201 ns/op          30.46 MB/s         441 B/op          6 allocs/op
BenchmarkPut/256B-8               300000              6066 ns/op          42.20 MB/s         571 B/op          6 allocs/op
BenchmarkPut/512B-8               200000              8382 ns/op          61.08 MB/s         861 B/op          6 allocs/op
BenchmarkPut/1K-8                 100000             12302 ns/op          83.23 MB/s        1443 B/op          6 allocs/op
BenchmarkPut/2K-8                 100000             21461 ns/op          95.43 MB/s        2606 B/op          6 allocs/op
BenchmarkPut/4K-8                  30000             39842 ns/op         102.81 MB/s        5187 B/op          6 allocs/op
BenchmarkPut/8K-8                  20000             89331 ns/op          91.70 MB/s        9843 B/op          6 allocs/op
BenchmarkPut/16K-8                 10000            142496 ns/op         114.98 MB/s       18881 B/op          6 allocs/op
BenchmarkPut/32K-8                  5000            284729 ns/op         115.08 MB/s       41580 B/op          7 allocs/op
BenchmarkScan-8                  1000000              1569 ns/op             408 B/op         16 allocs/op
```

Comparing both (notice that `WithPool` tests in the new version don't exist because isn't a configuration anymore):
```
benchmark               old ns/op     new ns/op     delta
BenchmarkGet/128B-8     1200          937           -21.92%
BenchmarkGet/256B-8     1222          889           -27.25%
BenchmarkGet/512B-8     1411          1051          -25.51%
BenchmarkGet/1K-8       1775          1249          -29.63%
BenchmarkGet/2K-8       2499          1629          -34.81%
BenchmarkGet/4K-8       3883          2492          -35.82%
BenchmarkGet/8K-8       6588          3844          -41.65%
BenchmarkGet/16K-8      11548         6965          -39.69%
BenchmarkGet/32K-8      22002         14140         -35.73%
BenchmarkPut/128B-8     4201          4016          -4.40%
BenchmarkPut/256B-8     6066          5209          -14.13%
BenchmarkPut/1K-8       12302         12305         +0.02%
BenchmarkPut/2K-8       21461         19231         -10.39%
BenchmarkPut/4K-8       39842         37505         -5.87%
BenchmarkPut/8K-8       89331         69770         -21.90%
BenchmarkPut/16K-8      142496        137953        -3.19%
BenchmarkPut/32K-8      284729        264327        -7.17%
BenchmarkScan-8         1569          1547          -1.40%

benchmark               old MB/s     new MB/s     speedup
BenchmarkGet/128B-8     106.58       136.54       1.28x
BenchmarkGet/256B-8     209.34       287.73       1.37x
BenchmarkGet/512B-8     362.80       486.93       1.34x
BenchmarkGet/1K-8       576.72       819.26       1.42x
BenchmarkGet/2K-8       819.35       1256.74      1.53x
BenchmarkGet/4K-8       1054.79      1643.57      1.56x
BenchmarkGet/8K-8       1243.32      2130.90      1.71x
BenchmarkGet/16K-8      1418.71      2352.30      1.66x
BenchmarkGet/32K-8      1489.26      2317.39      1.56x
BenchmarkPut/128B-8     30.46        31.87        1.05x
BenchmarkPut/256B-8     42.20        49.14        1.16x
BenchmarkPut/1K-8       83.23        83.21        1.00x
BenchmarkPut/2K-8       95.43        106.49       1.12x
BenchmarkPut/4K-8       102.81       109.21       1.06x
BenchmarkPut/8K-8       91.70        117.41       1.28x
BenchmarkPut/16K-8      114.98       118.76       1.03x
BenchmarkPut/32K-8      115.08       123.97       1.08x

benchmark               old allocs     new allocs     delta
BenchmarkGet/128B-8     4              1              -75.00%
BenchmarkGet/256B-8     4              1              -75.00%
BenchmarkGet/512B-8     4              1              -75.00%
BenchmarkGet/1K-8       4              1              -75.00%
BenchmarkGet/2K-8       4              1              -75.00%
BenchmarkGet/4K-8       4              1              -75.00%
BenchmarkGet/8K-8       4              1              -75.00%
BenchmarkGet/16K-8      4              1              -75.00%
BenchmarkGet/32K-8      4              1              -75.00%
BenchmarkPut/128B-8     6              6              +0.00%
BenchmarkPut/256B-8     6              6              +0.00%
BenchmarkPut/1K-8       6              6              +0.00%
BenchmarkPut/2K-8       6              6              +0.00%
BenchmarkPut/4K-8       6              6              +0.00%
BenchmarkPut/8K-8       6              6              +0.00%
BenchmarkPut/16K-8      6              6              +0.00%
BenchmarkPut/32K-8      7              7              +0.00%
BenchmarkScan-8         16             16             +0.00%

benchmark               old bytes     new bytes     delta
BenchmarkGet/128B-8     387           160           -58.66%
BenchmarkGet/256B-8     643           288           -55.21%
BenchmarkGet/512B-8     1187          576           -51.47%
BenchmarkGet/1K-8       2275          1152          -49.36%
BenchmarkGet/2K-8       4451          2304          -48.24%
BenchmarkGet/4K-8       9059          4864          -46.31%
BenchmarkGet/8K-8       17763         9472          -46.68%
BenchmarkGet/16K-8      34915         18432         -47.21%
BenchmarkGet/32K-8      73827         40960         -44.52%
BenchmarkPut/128B-8     441           201           -54.42%
BenchmarkPut/256B-8     571           202           -64.62%
BenchmarkPut/1K-8       1443          210           -85.45%
BenchmarkPut/2K-8       2606          221           -91.52%
BenchmarkPut/4K-8       5187          241           -95.35%
BenchmarkPut/8K-8       9843          288           -97.07%
BenchmarkPut/16K-8      18881         364           -98.07%
BenchmarkPut/32K-8      41580         531           -98.72%
BenchmarkScan-8         408           408           +0.00%
```

I'll tidy up these changes when merging with the other PR that makes keys `[]byte`.
When I create the PR, I would like second eyes to check this change and to validate this result. :100:

**Comment by prologic** [comment]
Holly crap that's impressive! Well done! Also Go Golang! Who knew you could beat the performance of Protobuf by hand-crafting your own serialization ;)

I'll definitely for sure verify these results on my shitty little 11" Macbook 🙃

This work also makes #46 more valuable now because by itself it doesn't achieve much in performance; only the convenience of using byte slices.

@awnumar When @jsign puts up the PR I'd love you to verify the results too in your use-case :) -- Although do you need us to write a export/import tool since we're changing the on-disk formatj?

**Comment by prologic** [comment]
@jsign You might be interested in [bitcask-bench](https://github.com/prologic/bitcask-bench) which I forked a while ago and added bitcask. I updated the READE with the latest results from v0.3.1 on my 11" Macbook.

**Comment by jsign** [comment]
Cool. I'll check those benchmarks after we merge  #49 & #46  that improve performance to see the final result.

**Comment by prologic** [comment]
Closed via #52
